### DISCUSSION TITLE
Custom VuforiaTrackables

### DISCUSSION THREAD

**Discussion by baqwas** [FIRST-Tech-Challenge/FtcRobotController#512] [Ideas] [answered]
Not sure if this is the right place to inquire but at least I am assured of an authoritative technical response based on past experience.

Using Vuforia can get more interesting if custom images can be loaded. Are there any plans to offer any "asset building" utility for this purpose? Perhaps Vuforia could be nudged into enabling custom _trackables_ that should surely lead to more productive use of its libraries by FTC teams?

**Comment by Windwoes** [comment, marked as answer]
The Vuforia support in the SDK should already support custom trackables simply by changing [which asset is loaded](https://github.com/FIRST-Tech-Challenge/FtcRobotController/blob/master/FtcRobotController/src/main/java/org/firstinspires/ftc/robotcontroller/external/samples/ConceptVuforiaFieldNavigation.java#L140).
You can use the [Vuforia Target Manager](https://developer.vuforia.com/target-manager) to create [Image Target](https://library.vuforia.com/objects/image-targets) assets which you can then include in your Android Studio project.

**Reply by baqwas** [reply]
Wish I had paid attention to this method much earlier in the season. This is exactly what I would like our team to explore. Anyhow, good to give our select few ML team members another angle into building models. Thanks to @texasdiaz and others we had success with the COCO models and this will be another approach to evaluate. Thanks again!

**Comment by gearsincorg** [comment]
I will follow up @Windwoes answer with the additional information that Vuforia is able to specifically track an image wrapped around a cone.  The standard Target Manager feature of the Vuforia Developer site will let you upload an image to be wrapped around a cone or cylinder.  

Our team used this method for PowerPlay.  It has proven quite sucessfull.  

![SmallCone](https://user-images.githubusercontent.com/14909461/205963896-a3c3b62c-7bd3-4025-8269-0af553248610.jpg)

Vuforia provides the cone rotation angle, and we use this to determine which image is facing the robot,

**Reply by baqwas** [reply]
@gearsincorg,

Thanks for the heads up. I was wondering why Cone had been singled out as a _type_.  I had given up on direct use of Vuforia especially after not having much luck engaging more experienced coaches/mentors in North Texas (DFW). Your suggestions will help me to make a U-turn and get our team back on Vuforia.

Kind regards.

**Reply by a-gt** [reply]
That sounds amazing, and our team will try it, do you think you could send some sample code on how to detect to come and it's rotation using vuforia.

**Reply by gearsincorg** [reply]
We based our code on theconceptVuforiaDriveToTargetWebcam sample code.

I've attached a vision class that we used to determine the cone orientation.   
This is a guide, not a complete solution....  Research and Experimentation are the best way to understand what is hapenning.

[GFORCE_Image.java.txt](https://github.com/FIRST-Tech-Challenge/FtcRobotController/files/10212809/GFORCE_Image.java.txt)

**Reply by a-gt** [reply]
Ok, thanks. I know what you mean by experimentation, I just wasn't sure how to implement it. Thanks a lot.
